// node-jira/lib/Login.js
/**
 * @author <a href="mailto:stefanmayer13@gmail.com">Stefan Mayer</a>
 */
'use strict';

exports.__esModule = true;
exports.loginRx = loginRx;
exports.login = login;
var https = require('https');
var Rx = require('rx');
var Boom = require('boom');

function loginRx(username, password) {
    var _this = this;

    return Rx.Observable.create(function (observer) {
        if (!username || !password) {
            observer.onError(Boom.badRequest('Missing username and/or password'));
            observer.onCompleted();
            return;
        }

        var postData = JSON.stringify({
            username: username,
            password: password
        });

        var url = '/rest/auth/1/session';

        var options = {
            hostname: _this.hostname,
            port: _this.port,
            path: url,
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(postData)
            }
        };

        _this.logger.info('Handling login to Jira', username);

        var req = https.request(options, function (res) {
            if (res.statusCode === 401) {
                return observer.onError(Boom.unauthorized('Username or password wrong'));
            } else if (res.statusCode === 403) {
                return observer.onError(Boom.forbidden('CAPTCHA required'));
            } else if (res.statusCode !== 200) {
                return observer.onError(Boom.create(res.statusCode, res.statusMessag));
            }

            observer.onNext({
                'setCookie': res.headers['set-cookie']
            });

            res.setEncoding('utf8');
            res.on('data', function (data) {
                observer.onNext(data);
            });
            res.on('end', function () {
                observer.onCompleted();
            });
        });

        req.on('error', function (e) {
            _this.logger.error('problem with request: ' + e.message);
            observer.onError(Boom.badGateway(e));
            observer.onCompleted();
        });
        req.write(postData);

        req.end();
    }).reduce(function (prev, current) {
        if (current.setCookie) {
            prev.setCookie = current.setCookie;
        } else {
            prev.data += current;
        }
        return prev;
    }, { data: '' }).map(function (data) {
        return {
            cookie: new Buffer(JSON.stringify(data.setCookie)).toString('base64'),
            data: JSON.parse(data.data)
        };
    });
}

function login(username, password) {
    return this.loginRx(username, password).toPromise();
}
